primitives/utility/src/lib.rs

```diff
@@ -120,11 +120,8 @@ impl<
 
 		let weight = Weight::from_ref_time(weight);
 
-		// We take the very first multiasset from payment
-		let multiassets: MultiAssets = payment.clone().into();
-
-		// Take the first multiasset from the selected MultiAssets
-		let first = multiassets.get(0).ok_or(XcmError::AssetNotFound)?;
+		// Take the very first multiasset from from payment
+		let first = payment.assets_iter().next().ok_or(XcmError::AssetNotFound)?;
```

Comment:
@joepetrowski 
I am just thinking, if it would be better here iterate and find "first.is_sufficient" maybe

Reply:
hmm, also , to remove `MultiAssets = payment.clone().into()` was not a good idea, there is sorting done in `impl From<Vec<MultiAsset>> for MultiAssets {`

Reply:
reverting `payment.clone().into();` it does that sorting `impl From<Vec<MultiAsset>> for MultiAssets {`

Reply:
now it takes "first" from sorted (probably by highest amount),
but it does not consider `is_sufficient` and/or `minimum_balance`,

so, the first asset could have higher amount then second,
but first asset could have also higher minimum_balance than second asset, so it couldnt be charged from first asset, but the second could be charged

Reply:
@joepetrowski 
do we want to change this "first selection" or let it be or just create following issue? wdyt?

at least, I am going to add here unit-test with multiple assets in payment

Reply:
@girazoki 
and please, wdyt about this "first selection" algo?

Reply:
Yeah, I like the iter approach. It would make it more user-friendly and make every effort to find an acceptable fee payment solution.

Reply:
multiassets object are always sorted I think (their creation involves sorting assets AFAIR). So we simply take the first.

There is no sufficiency indicator on a MultiAsset, so you probably will have to do db reads to know whether a multiasset is sufficient.

Besides, `BuyExecution` will always contain just one multiasset https://github.com/paritytech/polkadot/blob/53253de77f1649ecbb650df2ec3e29d0d1110a2d/xcm/src/v2/mod.rs#L564. The reason for `buy_weight` to receive `MUltiAssets` instead of `MultiAsset` as fee is because `try_take` from the holding register is defined in `MultiAsset`: https://github.com/paritytech/polkadot/blob/795b20c7150b04759ebddadd419e2f26fa382519/xcm/xcm-executor/src/lib.rs#L455

Reply:
@girazoki 
ok, thank you for the information,

yes, there have to be database hit for checking MultiAsset (is_suffiecient,min_balance), but this is already done in `AssetFeeAsExistentialDepositMultiplier -> BalanceConverter::to_asset_balance`.

and according to what you say, then maybe `let multiassets: MultiAssets = payment.clone().into();` the whole **clone/join_iterators/resorting** stuff is really not needed and `payment.fungible_assets_iter().next()` will do the same job with also less cloning of other stuff here, maybe without need to use Matcher here, so then maybe renaming `TakeFirstAssetTrader` to `TakeFirstFungibleAssetTrader`.

I understand, that TakeFirstAssetTrader is supposed to be as-much-as-possible abstract implementation, but I think, there is a little "space/tension" between API `payment: xcm_executor::Assets (MutliAssets)` vs. `BuyExecution { fees: MultiAsset`, and this "space/tension" could be possibly fixed by the improved "first selection" algo iterating multiassets. And based on the fact, that we are already doing database hit (is_suffiecient,min_balance) and "BuyExecution has only one asset", so there wont be any performance degradation, but opposite, without **clone/join_iterators/resorting** it could be just improved and on the other hand the algo will be more complete, maybe for future multiassets support for `BuyExecution(MutliAssets)`

we will come to this maybe later in the following issue: https://github.com/paritytech/cumulus/issues/1915